lib/BridgedRoom.js

```diff
@@ -0,0 +1,123 @@
+}
+
+BridgedRoom.prototype.matrixRoomId = function() {
+  return this._matrixRoomModel.getId();
+};
+
+BridgedRoom.prototype.gitterRoomName = function() {
+  return this._gitterRoomModel.getId();
+};
+
+BridgedRoom.prototype.joinAndStart = function() {
+  // We have to find out our own gitter user ID so we can ignore reflections of
+  // messages we sent
+  return this._bridge.getMyGitterUserId().then((gitterUserId) => {
+    this._gitterUserId = gitterUserId;
+
+    return this._gitter.rooms.join(this.gitterRoomName()).then((room) => {
+      this._gitterRoom = room;
+
+      var events = room.streaming().chatMessages();
+
+      events.on('chatMessages', (message) => {
+        if (!message.model.fromUser || message.model.fromUser.id == gitterUserId) {
+          // Ignore a reflection of my own messages
+          return;
+        }
+
+        if (message.operation !== 'create' && message.operation !== 'update') {
+          return;
+        }
+
+        this.onGitterMessage(message);
+      });
+
+      this._gitterRealtime.subscribe("/v1/rooms/" + room.id, (message) => {
+        this._bridge.getGitterUserById(message.userId).then((user) => {
+          if (user)
+            user.setRoomPresence(room.id, message.status == 'in');
+        });
+      });
+    });
+  });
+};
+
+BridgedRoom.prototype.stopAndLeave = function() {
+  this._gitterRoom.streaming().disconnect();
+
+  return this._gitterRoom.removeUser(this._gitterUserId);
+};
+
+function quotemeta(s) { return s.replace(/\W/g, '\\$&'); }
+
+BridgedRoom.prototype.onGitterMessage = function(message) {
+  var fromUser = message.model.fromUser;
+
+  console.log('gitter->' + this.gitterRoomName() + ' from ' + fromUser.username + ':', message.model.text)
+
+  this._bridge.mapGitterUser(fromUser).then((user) => {
+    return user.update(fromUser)
+      .catch((e) => {
+        console.log("Updating user failed:", e);
+        // ignore the failure and continue anyway
+      }).then(() => {
+        return user;
+      });
+  }).then((user) => {
+    var matrixMessage = {
+      msgtype: "m.text",
+      body: message.model.text,
+    };
+
+    // Pull out the HTML part of the body if it's not just plain text
+    if (message.model.html != message.model.text) {
+      matrixMessage["format"] = "org.matrix.custom.html";
+      matrixMessage["formatted_body"] = message.model.html;
+    }
+
+    if (message.model.status) {
+      matrixMessage["msgtype"] = "m.emote";
+
+      // Strip the leading @username mention from the body text
+      var userNameQuoted = quotemeta(fromUser.username);
+
+      matrixMessage["body"] =
+        matrixMessage["body"].replace(new RegExp("^@" + userNameQuoted + " "), "");
+
+      // TODO(paul): HTML is harder. Applying regexp mangling to an HTML string.
+      //   This is terrible. kegan - please help ;)
+      matrixMessage["formatted_body"] =
+        matrixMessage["formatted_body"].replace(new RegExp("^<span [^>]+>@" + userNameQuoted + "</span> "), "");
```

Comment:
It would be nice to see an example of the intended output of this. I'm assuming it is:
```
<span>@foo</span> bar baz     =>     bar baz
```

Assuming this is correct, is there really a problem here? You're just substringing the HTML in much the same way as the plain text. I don't see that as terrible if you know that this is the structure of the HTML. It becomes terrible if you need to introspect the HTML any more than this.